Add Watchdog timer and add support for NRF watchdog

xobs opened the pull request:
This patchset adds a new `wdt` module based on the WDT class provided by Micropython: http://docs.micropython.org/en/latest/library/machine.WDT.html

Note that on nrf, once the WDT is enabled it can not be disabled or have its duration modified without resetting the device.

xobs commented:
Initially I had it enabled for all NRF platforms, however that caused several platforms to overrun their flash. So I have since disabled all platforms except Simmel.

dhalbert commented:
> Initially I had it enabled for all NRF platforms, however that caused several platforms to overrun their flash. So I have since disabled all platforms except Simmel.

I'm surprised by this, since there is still a lot of unused flash on nRF52480, even with an internal filesystem. What's an example of a platform that overflowed? Maybe you were compiling with optimization off?

xobs commented:
Let me re-enable them for all platforms and re-build.

dhalbert commented:
@xobs ok, despite some missing builds (which are not boards), we can see that it didn't quite fit in the pca10100 build (by 80 bytes), so that's the (only) one to fix.

tannewt commented on ports/nrf/boards/common.template.ld:
```diff
@@ -70,7 +70,6 @@ SECTIONS
         _etext = .;        /* define a global symbol at end of code */
     } >FLASH_FIRMWARE
 
-    /*
```
What is extab?

xobs replied:
extab is the exception table, which is used in unwinding support on ARM. I thought maybe the exception system of circuit python used this table, which is overflowing enough that it is no longer fitting inline. I noticed that atmel-samd stores the extab in a similar location, so I simply emulated it here.

If circuit python isn't using this data at all, it's safe to discard this section. We could also try building with `-fno-unwind-tables` which should also prevent the generation of this section.

tannewt replied:
Huh, interesting. Is it needed for C++? Maybe we should leave it because we should add C++ library support at some point.

xobs replied:
It's used for printing out stack unwinding. For example, in Rust, if you discard this section it can't print a line number and function name when you `panic!()`.

If Circuit Python wants to start printing stack unwinding information for C or C++ modules, then this data should be saved. Otherwise, it should be safe to discard.

tannewt commented on shared-bindings/wdt/__init__.c:
```diff
@@ -0,0 +1,62 @@
+//|
+//| The `wdt` module provides support for a Watchdog Timer. This timer will reset the device
+//| if it hasn't been fed after a specified amount of time. This is useful to ensure the board
+//| has not crashed or locked up. You can enable thw watchdog timer using :class:`wdt.WDT`.
```
```suggestion
//| has not crashed or locked up. You can enable the watchdog timer using :class:`wdt.WDT`.
```

tannewt commented on shared-bindings/wdt/__init__.c:
```diff
@@ -0,0 +1,62 @@
+//|
+//| Example usage::
+//|
+//|     from machine import WDT
```
Update this.

tannewt commented on shared-bindings/wdt/WDT.c:
```diff
@@ -0,0 +1,128 @@
+
+const mp_obj_type_t wdt_wdt_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_WDT,
```
Please spell out the classname so it's easier to understand. `WatchDogTimer`

xobs replied:
I'll do that. However, note that the name `WDT` comes from micropython, so this change breaks some compatibility with them.

tannewt replied:
Totally ok.

tannewt commented on shared-bindings/wdt/__init__.c:
```diff
@@ -0,0 +1,62 @@
+//| The `wdt` module provides support for a Watchdog Timer. This timer will reset the device
+//| if it hasn't been fed after a specified amount of time. This is useful to ensure the board
+//| has not crashed or locked up. You can enable thw watchdog timer using :class:`wdt.WDT`.
+//| Note that the watchdog timer cannot be disabled once it has been enabled.
```
```suggestion
//| Note that the watchdog timer on some platforms cannot be disabled once it has been enabled.
```

tannewt commented on shared-bindings/wdt/WDT.c:
```diff
@@ -0,0 +1,128 @@
+//| class WDT:
+//|     """Watchdog Timer"""
+//|
+//|     def __init__(self, ):
```
Please add the params here and their docs. The timeout should be in float seconds to match `time.sleep`.

Please also elaborate that this will completely reset the microcontroller and not just Python.

tannewt commented on shared-bindings/wdt/WDT.c:
```diff
@@ -0,0 +1,128 @@
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+
+    if (wdt_singleton)
+        return wdt_singleton;
```
Is it possible to change the time and sleep settings? It'd be weird to pass them in to no effect. Maybe this should be a singleton on `supervisor` or `microcontroller` instead similar to `microcontroller.cpu`.

xobs replied:
I can place it there, but does that mean that we need to have a Watchdog on every platform? Or will it be like NVM, and will default to `mp_const_none_obj` if `CIRCUITPY_WATCHDOG` is `0`?

On NRF, it is not possible to disable the WDT, nor is it possible to adjust the settings -- the WDT registers are only unlocked after a reset.

tannewt replied:
I was thinking it'd be like nvm where it is None when CIRCUITPY_WATCHDOG = 0.

tannewt commented on shared-bindings/wdt/__init__.c:
```diff
@@ -0,0 +1,62 @@
+
+//| """Watchdog Timer
+//|
+//| The `wdt` module provides support for a Watchdog Timer. This timer will reset the device
```
I think `watchdog` would be a better module name because it's more verbose.

tannewt commented:
It'd be awesome to have the result of the watchdog be configurable. For testing, raising an exception rather than restarting would be very helpful. I imagine it's possible that libraries take a while to return to the main code.

xobs commented:
I put together a messy patch that implements what you requested. Documentation is unclear, but I wanted to make sure you were okay with the API before documenting everything again.

Example usage:

Simple usage:
```
wdt = microcontroller.watchdog.WatchDogTimer(5.0)
print("Sleeping for 3 seconds (shouldn't exit)")
time.sleep(3)
wdt.feed()
print("Sleeping for 3 seconds (also shouldn't exit)")
time.sleep(3)
print("Sleeping for 3 seconds without feeding WDT (should exit)")
time.sleep(3)
print("Shouldn't get here")
```

Using a `with` block:
```python
import microcontroller

with microcontroller.watchdog.WatchDogTimer(5.0) as wdt:
    print("Sleeping for 3 seconds (shouldn't exit)")
    time.sleep(3)
    wdt.feed()
    print("Sleeping for 3 seconds (also shouldn't exit)")
    time.sleep(3)
    print("Sleeping for 3 seconds without feeding WDT (should exit)")
    time.sleep(3)
    print("Shouldn't get here")
```

You can also add `hardware=True` to get the true Hardware watchdog timer.

---

For some reason, I'm unable to actually get this to work. It seems as though there's no way to set an exception in an interrupt routine, and while I'm setting `mp_pending_exception` this value is not immediately consulted.  As a result, "Shouldn't get here" is printed, although `time.sleep(3)` returns immediately.

It looks like the Python VM only checks for exceptions on loop constructs.  Additionally, when it sees an exception it calls `nlr_pop()`, but because we raised the exception in an interrupt handler there is no context to pop so it crashes.

tannewt commented on ports/nrf/common-hal/microcontroller/__init__.c:
```diff
@@ -117,6 +116,16 @@ const nvm_bytearray_obj_t common_hal_mcu_nvm_obj = {
+// The singleton nvm.WatchDogTimer object.
+const watchdog_obj_t common_hal_mcu_watchdog_obj = {
+    .base = {
+        .type = &watchdog_type,
```
I think you have one extra type here. This should be the watchdogtimer itself.

tannewt commented on py/modbuiltins.c:
```diff
@@ -719,6 +719,9 @@ STATIC const mp_rom_map_elem_t mp_module_builtins_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_IndexError), MP_ROM_PTR(&mp_type_IndexError) },
     { MP_ROM_QSTR(MP_QSTR_KeyboardInterrupt), MP_ROM_PTR(&mp_type_KeyboardInterrupt) },
     { MP_ROM_QSTR(MP_QSTR_ReloadException), MP_ROM_PTR(&mp_type_ReloadException) },
+    #if CIRCUITPY_WATCHDOG
```
This shouldn't be built-in. It can be a member of the `watchdog` module.

tannewt commented:
> I put together a messy patch that implements what you requested. Documentation is unclear, but I wanted to make sure you were okay with the API before documenting everything again.

Totally ok. Feel free to ping me on Discord when you get to your computer in the morning and we can video chat to sync up.

> 
> Example usage:
> 
> Simple usage:

I'd do:
```python
# By default the watchdog has infinite timeout and raises an exception on timeout.
wdt = microcontroller.watchdog 
# Now we set the timeout is 5 seconds and it will raise an exception. This may need to
# be implemented without the hardware WDT because it may only do real resets.
wdt.timeout = 5 
# This is where we start the real WDT to trigger a reset. Setting action to 
# anything else after will raise an exception (and likely lead to a WDT reset since it isn't fed 
# outside the VM.)
wdt.action = RESET 
print("Sleeping for 3 seconds (shouldn't exit)")
time.sleep(3)
wdt.feed()
print("Sleeping for 3 seconds (also shouldn't exit)")
time.sleep(3)
print("Sleeping for 3 seconds without feeding WDT (should exit)")
time.sleep(3)
print("Shouldn't get here")
```

With this API, the WatchdogTimer instance already exists. This is helpful because it's state will live across VMs.

Another thing to think about is the watchdog's relationship to safe mode. When off USB we probably want to boot up in normal mode after a watchdog. However, when on USB we may want to boot into safe mode or silently ignore action set to RESET if we started up on USB after a watchdog reset. If we do neither, we risk the user getting into a situation where they can't easily edit code.py due to a reset loop. There is a magic reset press to enter safe mode but most folks probably don't know about it.

> Using a `with` block:
> 

Why support a `with` block? It seems unhelpful because the nRF WDT can't be disabled.

> For some reason, I'm unable to actually get this to work. It seems as though there's no way to set an exception in an interrupt routine, and while I'm setting `mp_pending_exception` this value is not immediately consulted. As a result, "Shouldn't get here" is printed, although `time.sleep(3)` returns immediately.
> 
> It looks like the Python VM only checks for exceptions on loop constructs. Additionally, when it sees an exception it calls `nlr_pop()`, but because we raised the exception in an interrupt handler there is no context to pop so it crashes.

You should be able to set the pending exception from an ISR. We won't want to raise from there though. Maybe delay_ms needs to mp_raise the pending_exception? Does it work as is if you use a for loop to delay?

xobs commented:
I've implemented the new API.  Additionally, it now boots into safe mode if it's connected to USB and is reset due to a watchdog timeout.

I will work to disentangle the watchdog exception stuff from the vm.

xobs commented:
This patchset got big. But it was refactored twice.  It contains the following features:

* The watchdog is a singleton under `microcontroller`
* Setting `microcontroller.watchdog.mode = watchdog.WatchDogMode.RAISE` works
* If `watchdog.WatchDogMode.RESET` causes a reset, and the board is plugged in to USB, then it enters safe mode
* When the watchdog timer hits, it calls `reset_cpu()`
* Calling `microcontroller.reset()` now goes through `reset_cpu()`
* The number of system ticks is saved in uninitialized memory, so the clock is *mostly* correct after a reset

Simmel exceeded flash by 500 bytes, so I shrank the filesystem by 4096 bytes.

pca10100 has now exceeded its flash. I could disable watchdog on this platform.

I'm not sure why xtensa failed to build. It looks like a configuration error on the server.

dhalbert commented:
To get pca10100 to build (it's overflowing), change the options list in `mpconfigboard.mk` to this:
```
CIRCUITPY_AUDIOMP3 = 0
CIRCUITPY_BITBANGIO = 0
CIRCUITPY_BUSIO = 1
CIRCUITPY_COUNTIO = 0
CIRCUITPY_DISPLAYIO = 0
CIRCUITPY_FRAMEBUFFERIO = 0
CIRCUITPY_FREQUENCYIO = 0
CIRCUITPY_I2CSLAVE = 0
CIRCUITPY_NEOPIXEL_WRITE = 0
CIRCUITPY_NVM = 0
CIRCUITPY_PIXELBUF  = 0
CIRCUITPY_RGBMATRIX = 0
CIRCUITPY_ROTARYIO = 0
CIRCUITPY_RTC = 1
CIRCUITPY_TOUCHIO = 0
CIRCUITPY_ULAB = 0

SUPEROPT_GC = 0
```
This turns off several more modules and disables some extra optimization for GC. An alternative would be to shrink the internal filesystem size.

I would have a made a commit for this, but your fork isn't set to allow external changes via the PR mechanism.

dhalbert commented:
By the way, it appears from the nRF52833 errata and datsheet that it does not suffer from the nRF52840 SPIM3 hardware bug. I'm not sure if you're avoiding SPIM3 for simmel or not, but we should not need to allocate the special 8KB SPIM3 RAM buffer on any nRF52833 board and also not include the special case handling of SPIM3.

xobs commented:
@dhalbert Thanks for the heads-up! I am indeed disabling SPIM3 because the 8 kB penalty is pretty hefty on this 128 kB part. I'll need to do some special casing to only enable the workaround for NRF52840 parts. I'll do that in a separate patch.

I've made the adjustments you requested, and it's building now.

xobs commented:
I discovered an issue where the Bluetooth system couldn't be re-initialzied after the hardware watchdog was set, due to the hardware watchdog timer using IRQ priority 1.

dhalbert commented:
@tannewt The only build failures here are ESP32S2; is this fixed now?
@xobs Assuming ESP32S2  failures are a red herring, this just needs a merge from upstream and `make translate` to work, and then it should be fine.

tannewt commented on ports/nrf/common-hal/watchdog/WatchDogTimer.c:
```diff
@@ -0,0 +1,316 @@
+    reset_cpu();
+}
+
+//|     def feed(self):
+//|         """Feed the watchdog timer. This must be called regularly, otherwise
+//|         the timer will expire."""
+//|         ...
+//|
+STATIC mp_obj_t watchdog_watchdogtimer_feed(mp_obj_t self_in) {
```
These docs and the argument parsing should be in `shared-bindings` so that the API is shared amongst all ports. The port specific code should be called via a `common_hal_` routine.

tannewt commented on ports/nrf/common-hal/watchdog/WatchDogTimer.c:
```diff
@@ -0,0 +1,316 @@
+}
+MP_DEFINE_CONST_FUN_OBJ_2(watchdog_watchdogtimer_set_mode_obj, watchdog_watchdogtimer_obj_set_mode);
+
+const mp_obj_property_t watchdog_watchdogtimer_mode_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&watchdog_watchdogtimer_get_mode_obj,
+              (mp_obj_t)&watchdog_watchdogtimer_set_mode_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+STATIC const mp_rom_map_elem_t watchdog_watchdogtimer_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_feed), MP_ROM_PTR(&watchdog_watchdogtimer_feed_obj) },
+    { MP_ROM_QSTR(MP_QSTR_deinit), MP_ROM_PTR(&watchdog_watchdogtimer_deinit_obj) },
+    { MP_ROM_QSTR(MP_QSTR_timeout), MP_ROM_PTR(&watchdog_watchdogtimer_timeout_obj) },
+    { MP_ROM_QSTR(MP_QSTR_mode), MP_ROM_PTR(&watchdog_watchdogtimer_mode_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(watchdog_watchdogtimer_locals_dict, watchdog_watchdogtimer_locals_dict_table);
+
+const mp_obj_type_t watchdog_watchdogtimer_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_WatchDogTimer,
+    // .make_new = watchdog_watchdogtimer_make_new,
+    .locals_dict = (mp_obj_dict_t*)&watchdog_watchdogtimer_locals_dict,
+};
```
This all should be in `shared-bindings` as well.

tannewt commented on shared-bindings/watchdog/WatchDogMode.c:
```diff
@@ -0,0 +1,88 @@
+//|     def __init__(self, ):
+//|         """Enum-like class to define the run mode of the watchdog timer."""
+//|
+//|     NONE: Any = ...
+//|     """Take no action if the watchdog timer expires.
```
I don't think you need a NONE enum value. You can just use Python's `None`.

tannewt commented on shared-bindings/watchdog/__init__.c:
```diff
@@ -0,0 +1,84 @@
+//|     w.feed()"""
+//|
+
+const mp_obj_type_t mp_type_WatchDogTimeout = {
```
I wonder if we want this to be available in `watchdog` so people can try to catch it with `except`. What do you think?

xobs replied:
By this, you mean exposing this type as `watchdog.WatchDogTimeout`?

xobs replied:
This now works:

```python
import microcontroller
import watchdog
import time

wdt = microcontroller.watchdog
wdt.timeout = 5

while True:
    wdt.mode = watchdog.WatchDogMode.RAISE
    print("Starting loop -- should exit after five seconds")
    try:
        while True:
            time.sleep(10)
            # pass # This also works for a spinloop
    except watchdog.WatchDogTimeout as e:
        print("Watchdog Expired (PASS)")
    except Exception as e:
        print("Other exception (FAIL)")
print("Exited loop")
```

This prints out:

```
Starting loop -- should exit after five seconds
Watchdog Expired (PASS)
Starting loop -- should exit after five seconds
Watchdog Expired (PASS)
Starting loop -- should exit after five seconds
Watchdog Expired (PASS)
```

tannewt commented:
> @tannewt The only build failures here are ESP32S2; is this fixed now?

Ya, ESP32S2 should be happy again thanks to Jeff.

> @xobs Assuming ESP32S2 failures are a red herring, this just needs a merge from upstream and `make translate` to work, and then it should be fine.

xobs commented:
I've moved most everything into `shared-bindings`, rebased on current master, and exposed the exception under the `watchdog` module. Additionally, most checks now are performed inside `shared-bindings` with the exception of some platform-specific range checks.

xobs commented:
Thanks, I merged the PR.

`simmel-project` is an organization and not a user, and apparently you can't allow maintainer edits if it's an organization: https://github.com/isaacs/github/issues/1681

tannewt commented:
Welp, we merged more translations and I can't merge for you. Mind adding me as a contributor to the Simmel repo?

xobs commented:
Sure, I added you as a maintainer.
